gather.go

```diff
@@ -146,48 +147,64 @@ func (a *Agent) gatherCandidates() {
 		a.log.Warnf("Failed to run onCandidateHdlr task: %v\n", err)
 		return
 	}
-	a.gatheringState = GatheringStateComplete
 
+	if err := a.run(func(agent *Agent) {
+		a.gatheringState = GatheringStateComplete
+	}); err != nil {
+		a.log.Warnf("Failed to update gatheringState: %v\n", err)
+		return
+	}
 }
 
 func (a *Agent) gatherCandidatesLocal(networkTypes []NetworkType) {
 	localIPs := localInterfaces(networkTypes)
+
+	var wg sync.WaitGroup
+	wg.Add(len(localIPs) * len(supportedNetworks))
 	for _, ip := range localIPs {
 		for _, network := range supportedNetworks {
-			conn, err := listenUDP(int(a.portmax), int(a.portmin), network, &net.UDPAddr{IP: ip, Port: 0})
-			if err != nil {
-				a.log.Warnf("could not listen %s %s\n", network, ip)
-				continue
-			}
+			go func(network string, ip net.IP) {
```

Comment:
Possible to put the `wg.Done()` in a deferred?

Reply:
done.

Reply:
It seems defer wg.wait